app/models/notification.rb

```diff
@@ -14,6 +14,11 @@ class Notification < ApplicationRecord
     date_alert_due_date: 11
   }.freeze
 
+  REASON_TO_SCHEMA_PROPERTY = {
+    date_alert_start_date: "start_date",
+    date_alert_due_date: "due_date"
+  }.freeze
+
```

Comment:
I am unsure about the placement of this. It is more related to the representation in the API level than it is to the model itself.

There is a piece of code already [dealing with this sort of mapping](https://github.com/opf/openproject/blob/dev/lib/api/v3/notifications/details_factory.rb). It currently only caters to mapping the reason to the details but we could extend the interface to allow:

```
# Fetch the collection of details for a notification 
# (currently that would be an array with a single subclass of the API::V3::Values::PropertyRepresenter)
# It is the renamed #for method.
def details_for(notification)
...
end

# Fetch the collection of schemas for the provided notifications.
# It more or less does what is currently done within the NotificationCollectionRepresenter for mapping the
# notification reason to a ValueSchema.
def schemas_for(notifications)
...
end

# Returns the outward facing reason e.g. `dateAlert` as opposed to `date_alert_start_date`.
def reason_for(notification)
...
end
```

That way, we could also get rid of the rather simplistic [switch statement](https://github.com/opf/openproject/blob/dev/lib/api/v3/notifications/notification_representer.rb#L46-L51) I put in the nowhere of the representer.

Lastly, a renaming of the `DetailsFactory` would be required since it then does way more.

Reply:
@ulferts thanks for the suggestions!
I renamed the `DetailsFactory` to `PropertyFactory` as we are using it to provide property values for the notification representer, but I am open for naming suggestions.